### PR TITLE
Add RUN_RETURN to d_do_test and postprocess RUN_OUTPUT

### PR DESCRIPTION
**Pull request by maxhaton**
This is to support testing druntime with this tool.

### CODE REVIEW

**Comment by maxhaton** [comment]
@dkorpel @thewilsonator

File: compiler/test/tools/d_do_test.d [hunk 1]
```diff
@@ -103,6 +103,7 @@ struct TestArgs
     string   compileOutput;         /// `TEST_OUTPUT`: expected output of dmd
     string   compileOutputFile;     /// `TEST_OUTPUT_FILE`: file containing the expected `TEST_OUTPUT`
     string   runOutput;             /// `RUN_OUTPUT`: expected output of the compiled executable
+    ubyte    runReturn;             /// `RUN_RETURN` expected rc
```
Comment (thewilsonator) [inline compiler/test/tools/d_do_test.d:106]:
```diff
-    ubyte    runReturn;             /// `RUN_RETURN` expected rc
+    ubyte    runReturn;             /// `RUN_RETURN` expected exit code of the run executable
```

**Review by thewilsonator — APPROVED** [review]
I presume windows and posix exit codes are not the same, do we care about trying to harmonise them at all, or can we accomodate that with the infrastructure we already have?

This otherwise looks great!

**Comment by maxhaton** [comment]
> I presume windows and posix exit codes are not the same, do we care about trying to harmonise them at all, or can we accomodate that with the infrastructure we already have?
> 
> This otherwise looks great!

Potentially a good point, but i think this does what's needed for druntime tests (i.e. basically just 1 or 0)

**Comment by thewilsonator** [comment]
>  (i.e. basically just 1 or 0)

segfaults though? Or is that beyond the scope for now?

**Comment by ibuclaw** [comment]
> This is to support testing druntime with this tool.

Don't put druntime tests in `runnable/` please.

**Comment by thewilsonator** [comment]
> Don't put druntime tests in runnable/ please.

I think the plan is rather to use `d_do_test` to supplant the use of `Makefiles` in druntime, rather than druntime tests in `runnable/`.

**Comment by maxhaton** [comment]
> > This is to support testing druntime with this tool.
> 
> Don't put druntime tests in `runnable/` please.

The next patch adds support for a new folder, which when working will be moved somewhere else, but that involves moving the test runner so baby steps.

**Comment by maxhaton** [comment]
And yes we have way too many tests in runnable.

@ibuclaw does gdc use these druntime tests as-is?

**Comment by ibuclaw** [comment]
> And yes we have way too many tests in runnable.
> 
> @ibuclaw does gdc use these druntime tests as-is?

For things under druntime/tests I converted them all to Dejagnu.

Most require nothing special, but for example

https://github.com/gcc-mirror/gcc/blob/c131b1d5da412b9b0f0681fc5704cdc9b7cafb04/libphobos/testsuite/libphobos.cycles/cycles.exp#L21

https://github.com/gcc-mirror/gcc/blob/c131b1d5da412b9b0f0681fc5704cdc9b7cafb04/libphobos/testsuite/libphobos.cycles/mod1.d#L1

**Comment by maxhaton** [comment]
> > And yes we have way too many tests in runnable.
> > @ibuclaw does gdc use these druntime tests as-is?
> 
> For things under druntime/tests I converted them all to Dejagnu.
> 
> Most require nothing special, but for example
> 
> https://github.com/gcc-mirror/gcc/blob/c131b1d5da412b9b0f0681fc5704cdc9b7cafb04/libphobos/testsuite/libphobos.cycles/cycles.exp#L21
> 
> https://github.com/gcc-mirror/gcc/blob/c131b1d5da412b9b0f0681fc5704cdc9b7cafb04/libphobos/testsuite/libphobos.cycles/mod1.d#L1

Forewarning then that the main reason why I'm shuffling these around is so I can make changes to their output!

**Comment by maxhaton** [comment]
Changed comment, have rebased (god bless shithub...)

File: compiler/test/tools/d_do_test.d [hunk 2]
```diff
@@ -828,6 +829,17 @@ bool gatherTestParameters(ref TestArgs testArgs, string input_dir, string input_
 
     findOutputParameter(file, "RUN_OUTPUT", testArgs.runOutput, envData);
 
+    string dummy;
+    if (findTestParameter(envData, file, "RUN_RETURN", dummy))
+    {
+        import std.conv : to;
```
Comment (thewilsonator) [inline compiler/test/tools/d_do_test.d:835]: this file already imports `std.conv`

**Pull request merged**